piptools/repositories/pypi.py

```diff
@@ -99,8 +112,9 @@ def get_dependencies(self, ireq):
         dependencies (also InstallRequirements, but not necessarily pinned).
         They indicate the secondary dependencies for the given requirement.
         """
-        if not (ireq.editable or is_pinned_requirement(ireq)):
-            raise TypeError('Expected pinned or editable InstallRequirement, got {}'.format(ireq))
+
+        if ireq in self._get_dependencies_cache:
+            return self._get_dependencies_cache[ireq]
```

Comment:
IIRC it comes from https://github.com/blueyed/pip-tools/commit/82aec97a10d5af438eff86df1f3921bbf83ff029. It provides both a performance optimization, but is required actually, because otherwise `pip` would complain about an existing build dir.